pkg/pillar/diskmetrics/usage.go

```diff
@@ -241,3 +259,48 @@ func PersistUsageStat(log *base.LogObject) (*types.UsageStat, error) {
 	}
 	return usageStat, nil
 }
+
+// PathAndSize is returned by FindLargeFiles
+type PathAndSize struct {
+	Path string
+	Size int64
+}
+
+// FindLargeFiles walks a directory and reports all files larger than minSize
+// unless they are in an excluded (sub)directory
+func FindLargeFiles(root string, minSize int64, excludePaths []string) ([]PathAndSize, error) {
+	var list []PathAndSize
+	walkErr := filepath.WalkDir(filepath.Clean(root), func(path string, di fs.DirEntry, err error) error {
+		// if there is any problem with path we stop
+		if err != nil {
+			return err
+		}
+
+		// Part of excludePath?
+		for _, ex := range excludePaths {
+			if filepath.HasPrefix(path, ex) {
+				return filepath.SkipDir
+			}
+		}
+		// We don't report any directories
+		if di.IsDir() {
+			return nil
+		}
+
+		info, err := os.Stat(path)
```

Comment:
Btw. @andrewd-zededa made a change recently where he limited the number of Stat calls since it apparently can be expensive to call on each file: https://github.com/lf-edge/eve/pull/4065
I don't know if the same consideration needs to apply for FindLargeFiles as well..

Reply:
This is my only concern as well. `os.Stat` is expensive, as you have to read each file inode. Unfortunately, there is no easier way to do it that I can think of. The way ext4 is built, the directory entries include only the name of the entry and the type. If you want the size, you need to read the inode, for which you need `os.Stat()`.

As far as I can see #4065 just saves by not calling `os.Stat()` on files outside of a given path, which I think this does already.

If you _really_ want to get fancy, I have a library that knows how to read ext4 natively. It can load the entire inode table into memory, and you can use that to replace the `os.Stat()` for each one with an in-memory lookup. It won't save you on the directory tree reading, but it will save you all of the `os.Stat()` calls. It doesn't have a public interface for doing that, but I don't mind adding it.

I would suggest that we put this through its paces, including performance test. If that becomes an issue, worry about it then.

Reply:
We're only doing this in unknown directories in /persist - all of the known places are in the exclude paths.
This is to catch somebody manually e.g., putting huge tar/iso/img files in /persist/foo.iso or /persist/bar/foo.iso.
Thus the number of things we stat should be very small unless somebody manually dumps a huge directory tree manually into /persist.
Let me get a count on the number of calls.

Reply:
If it's that small, then don't worry too much about it. And definitely don't use that library.

Reply:
I checked a default freshly installed EVE device and there are only two files which get Stat'ed in this loop (/persist/SMART_details.json and /persist/SMART_details_previous.json). And the FindLargeFiles is called every 20 seconds. So this doesn't seem to be a risk for load.